# Pandas: Statistische Auswertungen

Pandas wurde als Werkzeug für die Datenanalyse entwickelt und ist somit dem Bereich *Data Science* zuzuordnen. Es ist damit natürlich vor allem für statistische Auswertungen geeignet und ist stark an die Möglichkeiten der Programmiersprache `R` angelehnt, die im Bereich der Statitik viel verwendet wird.

## Einlesen als DataFrame

Bisher haben wir immer unsere Daten direkt im Notebook erzeugt. In der Praxis ist es jedoch fast immer so, dass die Daten entweder aus einer Datei oder z.B. einer Datenbank geladen werden. Pandas bietet zu diesem Zweck eine Reihe von Hilfsfunktionen an, wie z.B. `read_csv()`, `read_table()`, `read_json()`, `read_hdf()`,  `read_xml()` usw. 

Die Daten für dieses Notebook lesen wir aus einer CSV (Comma Separated Values) Datei ein. Unsere Beispieldaten stammen von DASL (Data and Story Library: https://dasl.datadescription.com), einer Website, die Daten für Statistikkurse bereit stellt. Der verwendete Datensatz "Amazon books" beschreibt 325 Bücher in 12 Spalten: https://dasl.datadescription.com/datafile/amazon-books. Lesen wir die Daten einmal ein:

In [59]:
import pandas as pd

books = pd.read_csv('amazon-books.txt', sep='\t', encoding='UTF-8')

Die `read_csv()` Funktion erwartet als erstes Argument den Namen (ggf. mit Verzeichnispfad) der 
einzulesende Datei. Hier verwenden wir noch zwei weitere Argumente:

  * `sep=` gibt den Spaltentrenner der einzulesenden Datei an.
    Das 'C' in `CSV` steht ja eigentlich für 'Comma', die CSV Dateien können aber auch zum Beispiel
    durch ein Semikolon (`;`) oder ein Tabulaturzeichen (`\t`) getrennt werden. Streng genommen
    liegen unsere Daten also nicht als CSV, sondern als TSV (tabulator separated values) vor.
  * `encoding=` gibt die Kodierung an, in der die Daten in der Datei gespeichert sind.

Weitere oft nützliche Parameter der `read_csv()` Datei sind:

  * `header=0` teilt Pandas mit, dass die erste Zeile bereits Daten und nicht die Spaltennamen enthält
  * `names=`  Falls die Daten die Spaltennamen nicht in der ersten Zeile enthalten, können diese mit diesem Parameter vergeben werden
      z.B.: `names=['Name', 'Age']`.
  * `index_col=` Wird dieser Parameter nicht verwendet, erzeugt Pandas fortlaufende Zahlen als Index. `index_col=`
    erwartet den Namen einer Spalte als Wert. Diese Spalte wird dann für den Index verwendet.
  * `nrows=`  ist dann nützlich, wenn nicht alle Zeilen, sondern z.B. nur die ersten 100 eingelesen werden sollen. Das
    ist vor allem dann nützlich, wenn wir während der Entwicklungsphase uns nicht mit riesigen Datenmengen herumschlagen
    wollen.
  * `usecols=` wird verwendet, wenn wir nicht alle, sondern nur bestimmte Spalten aus der CSV-Datei
    verwenden wollen. Der Wert ist eine List der zu verwendenden Spaltennamen.
  * `dtype=` ist praktisch, wenn wir für eine oder mehrere Spalten einen bestimmten Datentyp erzwingen wollen. Fehlt dieser
    Parameter, versucht Pandas selbst einen geeigneten Typ zu wählen. Der Wert ist ein Dictionary, in dem der Key der Spaltenname ist
    und der Value der zu verwendende Typ. Beispiel:
    `pd.read_csv('mydata.csv', dtype={'Alter': 'uint8', 'salary': 'float'}`
    Eine Liste der möglichen Typen finden Sie hier: https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#basics-dtypes.

## Speichern eines DataFrames

Wir haben zwar unsere Daten noch nicht verändert, zeigen aber dennoch bereits hier, wie wir ein DataFrame wieder in eine CSV-Datei schreiben können. Pandas stellt dazu die Funktion `to_csv()` bereit. Wie beim Einlesen von Daten gibt es auch hier Funktionen für eine Reihe von Ausgabeformaten wie z.B. `to_excel()`,
`to_json()`usw.

Auch hier gibt es wieder einige Parameter, die verwendet werden können. Beispielsweise kann man mit `columns=[]` die Namen von Spalten festlegen, die in die Datei geschrieben werden sollen. Wir setzen hier das Semicolon als Trennzeichen und wieder 'UTF-8' als Encoding.

In [60]:
books.to_csv('../output/clean_books,csv', sep=';', encoding='UTF-8')

## Überblick über die Daten

Ehe wir etwas Sinnvolles mit den Daten anfangen können, müssen wir sie einmal verstehen. Pandas bietet hier einige Möglickeiten, die wir bereits kennen gelernt habe. Zur Auffrischung wenden wir sie hier auf echte, uns noch nicht bekannte Daten an.

### Ausschnitt anzeigen

Die `head()` Methode zeigt die ersten 5 Zeilen. Alternativ gibt es noch `tail()`(gibt die letzten 5 Zeilen aus) und `sample()` (gib eine zufällige Zeile aus).

Alle Methoden akzeptieren auch eine Zahl, über die die Zahl der ausgegebenen Zeilen gesteuert werden kann.

In [61]:
books.head(3)

,Title,Author,List Price,Amazon Price,Hard/ Paper,NumPages,Publisher,Pub year,ISBN-10,Height,Width,Thick,Weight (oz)
0,"1,001 Facts that Will Scare the S#*t Out of Yo...",Cary McNeal,12.95,5.18,P,304.0,Adams Media,2010.0,1605506249,7.8,5.5,0.8,11.2
1,21: Bringing Down the House - Movie Tie-In: Th...,Ben Mezrich,15.00,10.20,P,273.0,Free Press,2008.0,1416564195,8.4,5.5,0.7,7.2
2,100 Best-Loved Poems (Dover Thrift Editions),Smith,1.50,1.50,P,96.0,Dover Publications,1995.0,486285537,8.3,5.2,0.3,4.0


### Zahl der Zeilen und Spalten

`shape` enthält ein Tupel: Zahl der Zeilen und Zahl der Spalten. Bin ich nur an der Zahl der Spalten interessiert kann ich so drauf zugreifen: `shape[0]`.

In [62]:
books.shape  # Show number of lines and columns

(325, 13)

### Liste der Spaltennamen

`columns` enthält die Spaltennamen als Liste. Die Reihenfolge entspricht der Reihenfolge im DataFrame.

In [63]:
books.columns  # give me list with all column names

Index(['Title', 'Author', 'List Price', 'Amazon Price', 'Hard/ Paper',
       'NumPages', 'Publisher', 'Pub year', 'ISBN-10', 'Height', 'Width',
       'Thick', 'Weight (oz)'],
      dtype='object')

### Datentypen der Spalten

`dtypes` listet die Spaltennamen und die in den einzelnen Spalten verwendeten Datentypen auf.

In [64]:
books.dtypes

Title            object
Author           object
List Price      float64
Amazon Price    float64
Hard/ Paper      object
NumPages        float64
Publisher        object
Pub year        float64
ISBN-10          object
Height          float64
Width           float64
Thick           float64
Weight (oz)     float64
dtype: object

In [65]:
books['Pub year']

0      2010.0
1      2008.0
2      1995.0
3      2008.0
4      2011.0
        ...  
320    2004.0
321    2011.0
322    2005.0
323    2010.0
324    2011.0
Name: Pub year, Length: 325, dtype: float64

## Daten beim Einlesen adaptieren

Hier noch ein Beispiel, in dem wir beim `read_csv()` die eingelesenen Daten manipulieren. 

  - Wir setzen die Spaltennamen durch unsere eigenen (kürzeren) Spaltennamen
  - Wir setzen den Datentyp von `pages` und 'year' auf `Int16', Die explizite Angabe eines anderen Datentyps würde auch auch für weitere Spalten wie 'price' oder die Maße (width, heigh,m weight) Sinn machen. Allerdings stehen wir hier vor dem Problem,
    dass die meisten Datentypen (noch?) nicht *nullable* sind. Deshalb würde es beim Einlesen Fehler geben, wenn ein Wert fehlt.

Das hier erzeugte DataFrame dient nur zur Demonstartion der Möglichkeiten. In der Folge werden wir das zuerst erzeugte DataFrame mit den originalen Spaltennamen und Default-Datentypen verwenden.

In [66]:
column_names = ['title','author','listprice','price','hardcover','pages',
                'publisher','year','isbn','height', 'width','thick','weight']
column_dtype = {'year':'Int16', 'pages': 'Int16'}

amazon_books = pd.read_csv('amazon-books.txt',sep='\t',encoding='UTF-8',header = 0, 
                           names = column_names, dtype = column_dtype)

print(f'\nDatentyp der Spalten\n\n{amazon_books.dtypes}\n')


Datentyp der Spalten

title         object
author        object
listprice    float64
price        float64
hardcover     object
pages          Int16
publisher     object
year           Int16
isbn          object
height       float64
width        float64
thick        float64
weight       float64
dtype: object



## Daten vorbereiten

Ehe wir die Daten auswerten, müssen wir sie in einen Zustand bringen, die eine sinnvolle Auswertung erst ermöglicht. Dazu gehört vor allem die Überprüfung auf fehlende Werte.

### Überprüfung auf fehlende Werte

Wir haben ja bereits die Methoden `isna()`und `isnull()` kennengelernt. Prüfen wir einmal, ob es überall eine Seitenangabe gibt:

In [67]:
books.NumPages.isna()

0      False
1      False
2      False
3      False
4      False
       ...  
320    False
321    False
322    False
323    False
324    False
Name: NumPages, Length: 325, dtype: bool

Obwohl wir mit 325 Büchern einen sehr kleinen Datensatz haben, ist diese Ausgabenicht mehr wirklich übersichtlich. Daher zählen wir die auf `True` stehenden Ergebnisse aus:

In [68]:
print(books.NumPages.isna().sum())

2


`2` bedeutet hier, dass die Spalte zwei Werte enthält, die nicht als Zahl interpretierbar sind. Wie finden wir nun heraus, welche das sind?

In [69]:
books[books.NumPages.isna()]

,Title,Author,List Price,Amazon Price,Hard/ Paper,NumPages,Publisher,Pub year,ISBN-10,Height,Width,Thick,Weight (oz)
8,A Dance With Dragons,"Martin RR, George",35.0,18.81,H,NaN,Bantam,2011.0,553801473,9.6,6.5,2.1,NaN
76,Fences,August Wislon,21.1,16.46,H,NaN,Perfection Leraning,1986.0,780707508,9.0,6.0,1.0,22.4


`isna()` können wir auch auf einen ganzen DataFrame ausführen und so alle Spalten finden, in denen ein `NaN` Wert vorkommt:

In [70]:
books.isna().sum()

Title           0
Author          1
List Price      1
Amazon Price    0
Hard/ Paper     0
NumPages        2
Publisher       1
Pub year        1
ISBN-10         0
Height          4
Width           5
Thick           1
Weight (oz)     9
dtype: int64

Wie man das Problemmit den fehlenden Werten löst, ist nicht verallgemeinerbar. Je nach Datenlage und Auswertungsabsicht könnte man die `NaN` Werte durch einen Defaultwert ersetzen, bei kleinen Datensammlungen vielleicht nachrecherchieren usw. Die einfachste Lösung, insbesondere, wenn man einen großen Datenbestand hat, liegt in der Entfernung solcher Zeilen:

In [71]:
clean_books = books.dropna()
clean_books.isna().sum()

Title           0
Author          0
List Price      0
Amazon Price    0
Hard/ Paper     0
NumPages        0
Publisher       0
Pub year        0
ISBN-10         0
Height          0
Width           0
Thick           0
Weight (oz)     0
dtype: int64

## Statistische Kennzahlen

### Überblick mit describe()

Die Methode `describe()` generiert folgende statistische Kennzahlen für numerische Daten:

   * `count` ist die Zahl der (gültigen) Werte (Zeilen)
   * `mean` ist das arithmetische Mittel
   * `std` ist die Standardabweichung (durchschnittliche Abweichung vom Mittelwert)
   * `min` ist das Minimum, also der kleinste vorkommende Wert
   * `max` ist das Maximum, also der größe vorkommende Wert
   * `25` ist das erste Quartil. Also der Wert, der die Grenze zwischen dem ersten und zweiten Viertel der sortieren Werte darstellt.
   * `50` ist das zweite Quartil. Also der Wert, der die Grenze zwischen dem zweiten und dritten Viertel der sortieren Werte darstellt. Dies entspricht dem Median.
   * `75` ist das dritte Quartil. Also der Wert, der die Grenze zwischen dem dritten und vierten Viertel der sortieren Werte darstellt.

In [72]:
clean_books.describe()

,List Price,Amazon Price,NumPages,Pub year,Height,Width,Thick,Weight (oz)
count,310.000000,310.000000,310.000000,310.000000,310.000000,310.000000,310.000000,310.000000
mean,17.934516,12.587645,330.577419,2002.035484,8.130968,5.553548,0.890000,12.427742
std,13.287255,11.578859,159.017713,10.732796,0.868884,0.845416,0.355534,6.622459
min,1.500000,0.770000,24.000000,1936.000000,5.100000,4.100000,0.100000,1.200000
25%,13.425000,8.537500,208.000000,1998.000000,7.900000,5.200000,0.600000,7.800000
50%,15.000000,10.200000,320.000000,2005.000000,8.100000,5.300000,0.800000,11.200000
75%,18.000000,12.240000,403.750000,2010.000000,8.400000,5.800000,1.100000,16.000000
max,139.950000,139.950000,896.000000,2011.000000,12.100000,9.300000,2.000000,35.200000


Um die Sache übersichtlicher zu machen, runden wir die Werte auf zwei Nachkommastellen:

In [73]:
clean_books.describe().round(2)

,List Price,Amazon Price,NumPages,Pub year,Height,Width,Thick,Weight (oz)
count,310.00,310.00,310.00,310.00,310.00,310.00,310.00,310.00
mean,17.93,12.59,330.58,2002.04,8.13,5.55,0.89,12.43
std,13.29,11.58,159.02,10.73,0.87,0.85,0.36,6.62
min,1.50,0.77,24.00,1936.00,5.10,4.10,0.10,1.20
25%,13.42,8.54,208.00,1998.00,7.90,5.20,0.60,7.80
50%,15.00,10.20,320.00,2005.00,8.10,5.30,0.80,11.20
75%,18.00,12.24,403.75,2010.00,8.40,5.80,1.10,16.00
max,139.95,139.95,896.00,2011.00,12.10,9.30,2.00,35.20


Das funktioniert natürlich nur für numerische Werte. Wir können aber auch für nicht numerische Werte einen Überblick generieren:

In [74]:
clean_books['Author'].describe()

count                  310
unique                 243
top       Vladimir Nabokov
freq                     7
Name: Author, dtype: object

Das Ergebnis ist so zu lesen:

  * Es wurden 310 Werte gefunden, davon waren 243 unique. Oder anderes gesagt: es wurden 243 unterschiedliche Autoren gefunden.
  * Am öftesten kommt der Wert 'Vladimir Nabokov' vor, nämlich sieben Mal. Hier müssen wir ein wenig aufpassen, weil hier immer nur ein Wert ausgegeben wird. Hätten wir einen zweiten Autor mit 7 Büchern, würde dieser in dieser Ausgabe unterschlagen.

Wenn wir nicht-numerische Spalten in die durch `describe()` generierte Ausgabe aufnehmen wollen, müssen wir angeben, dass alle Spalten mit bestimmten Datentypen berücksichtigt werden sollen:

In [75]:
clean_books.describe(include=['object', 'int', float]).round(1)

,Title,Author,List Price,Amazon Price,Hard/ Paper,NumPages,Publisher,Pub year,ISBN-10,Height,Width,Thick,Weight (oz)
count,310,310,310.0,310.0,310,310.0,310,310.0,310,310.0,310.0,310.0,310.0
unique,300,243,NaN,NaN,2,NaN,153,NaN,309,NaN,NaN,NaN,NaN
top,The Great Gatsby,Vladimir Nabokov,NaN,NaN,P,NaN,Vintage,NaN,892967412,NaN,NaN,NaN,NaN
freq,3,7,NaN,NaN,232,NaN,37,NaN,2,NaN,NaN,NaN,NaN
mean,NaN,NaN,17.9,12.6,NaN,330.6,NaN,2002.0,NaN,8.1,5.6,0.9,12.4
std,NaN,NaN,13.3,11.6,NaN,159.0,NaN,10.7,NaN,0.9,0.8,0.4,6.6
min,NaN,NaN,1.5,0.8,NaN,24.0,NaN,1936.0,NaN,5.1,4.1,0.1,1.2
25%,NaN,NaN,13.4,8.5,NaN,208.0,NaN,1998.0,NaN,7.9,5.2,0.6,7.8
50%,NaN,NaN,15.0,10.2,NaN,320.0,NaN,2005.0,NaN,8.1,5.3,0.8,11.2
75%,NaN,NaN,18.0,12.2,NaN,403.8,NaN,2010.0,NaN,8.4,5.8,1.1,16.0


### Einzelne Kennzahlen

Ist man nur an einem bestimmten statistischen Wert interessiert, gibt es entsprechende Methoden.

#### Minimum und Maximum

`min()` und `max()` ermitteln den kleinsten bzw. größten Wert der Spalte.

In [76]:
print(clean_books.NumPages.min())
print(clean_books.NumPages.max())

24.0
896.0


#### Arithemtisches Mittel

Der Mittelwert wird über die Methode `mean()` ermittelt.

In [77]:
print(clean_books["List Price"].mean().round(2))
print(clean_books["Amazon Price"].mean().round(2))
print(clean_books["NumPages"].mean().round(2))

17.93
12.59
330.58


Den Mittelwert sollte man immer zusammen mit der Standardabweichung betrachen. Diese ist über die Methode `std()`verfügbar:

In [78]:
print(f"Listenpreis: {clean_books["List Price"].mean().round(2)} bei einer Standardabweichung von {clean_books["List Price"].std().round(2)}")
print(f"Amazon Preis: {clean_books["Amazon Price"].mean().round(2)} bei einer Standardabweichung von {clean_books["Amazon Price"].std().round(2)}")
print(f"Seitenzahl: {clean_books["NumPages"].mean().round(2)} bei einer Standardabweichung von {clean_books["NumPages"].std().round(2)}")

Listenpreis: 17.93 bei einer Standardabweichung von 13.29
Amazon Preis: 12.59 bei einer Standardabweichung von 11.58
Seitenzahl: 330.58 bei einer Standardabweichung von 159.02


#### Median

Der Median wird so ermittelt, dass man die einzelnen Werte sortiert und dann den (oder die) Werte ermittelt, die genau in der Mitte liegen. Deshalb entspricht der Median auch dem zweiten Quartil.

In [79]:
print(clean_books.NumPages.median().round(2))

320.0


#### Die Quantile
Die Quantile gehören zu den so genannten Lagemaßen, die vor allem in der Wahrscheinlichkeitsrechnung eine wichtige Rolle spielen. Dabei werden einfach die Daten sortiert und dann die Grenzen zwischen den Quantilen ermittelt. Dann lässt sich feststellen, in welchem Quantil ein Wert liegt. Das klassische Quantil ist das Quartil (Aufteilung in 4 Viertel), aber grundsätzlich kann man z.B. auch Quintile (Fünftel) oder Dezile (Zehntel) ermitteln. Die Aufteilung lässt sich in Pandas als Zahl zwischen 0 und 1 ausdrücken. `0.5` entspricht dabei dem Median, weil gleich viele Werte vor sowie nach dieser Grenze liegen. Probieren wir es aus:

In [80]:
print(clean_books.NumPages.quantile(0.5).round(2))

320.0


Wenn wir das erste Quartil haben wollen, müssen wir `0.25` angeben, `0.75` für das dritte Quartil usw. Für die Ermittlung der Grenze für das erste Dezil müssen wir `0.1` angeben:

In [81]:
print(clean_books.NumPages.quantile(0.1))

152.0


#### Modus

Der Modus (d.h. der häufigste Wert) lässt sich mit der Methode `mode()` ermitteln:

In [82]:
clean_books.NumPages.mode()

0    352.0
Name: NumPages, dtype: float64

#### Varianz

Die Varianz ist das Quadrat der mittleren Abweichungen. Die entsprechende Methode heißt in Pandas `var()`.

In [83]:
clean_books.NumPages.var()

np.float64(25286.633145422278)

### Werte auszählen

### Zahl der Werte

Um zu ermitteln, wie viele Werte in einer Spalte vorhanden sind, verwenden wir die Methode `count()`. Diese ignoriert fehlende Werte. Sie enspricht also nicht der Zahl der Zeilen, sondern ist der Zahl der Werte in einer Spalte, die nicht `None` sind.

In [84]:
print(books.Author.count())

324


Falls wir die Zahl distinkten Werte (also ohne mehrfache Vorkommen) brauchen, bietet sich die `nunique()` Methode an.

In [85]:
books.Author.nunique()

255

Schließlich lernen wir noch eine letzte Methode kennen: `value_counts()`. Diese gruppiert im Hintergrund nach Werten und zählt dann die Werte pro Gruppe. Auf diese Art können wir ganz einfach ermitteln, wie viele Bücher eines jeden Autors bzw. jeder Autorin vorhanden sind:

In [86]:
books.Author.value_counts()

Author
Vladimir Nabokov       7
Jodi Picoult           7
Lewis                  4
Murakami               4
Shel Silverstein       3
                      ..
Bruno Latour           1
Paula S. Rothenberg    1
Geralyn Lucas          1
Robert Tignor          1
Emily Bronte           1
Name: count, Length: 255, dtype: int64

## Lizenz

This notebook ist part of the course [Grundlagen der Programmierung](https://github.com/gvasold/gdp) held by [Gunter Vasold](https://online.uni-graz.at/kfu_online/wbForschungsportal.cbShowPortal?pPersonNr=51488) at Graz University 2017&thinsp;ff. 

<p>
    It is licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0">CC BY-NC-SA 4.0</a>
</p>

<table>
    <tr>
    <td>
        <img style="height:22px" 
             src="https://mirrors.creativecommons.org/presskit/icons/cc.svg?ref=chooser-v1"/></li>
    </td>
    <td>
    <img style="height:22px;"
         src="https://mirrors.creativecommons.org/presskit/icons/by.svg?ref=chooser-v1" /></li>
    </td>
    <td>
        <img style="height:22px;"
         src="https://mirrors.creativecommons.org/presskit/icons/nc.svg?ref=chooser-v1" /></li>
    </td>
    <td>
        <img style="height:22px;"
             src="https://mirrors.creativecommons.org/presskit/icons/sa.svg?ref=chooser-v1" /></li>
    </td>
</tr>
</table>